x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/snapshotlifecycle/SnapshotLifecycleServiceTests.java

```diff
@@ -0,0 +1,230 @@
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.xpack.core.scheduler.SchedulerEngine;
+import org.elasticsearch.xpack.core.watcher.watch.ClockMock;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+
+import static org.hamcrest.Matchers.containsInAnyOrder;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.greaterThan;
+
+public class SnapshotLifecycleServiceTests extends ESTestCase {
+
+    public void testGetJobId() {
+        String id = randomAlphaOfLengthBetween(1, 10) + (randomBoolean() ? "" : randomLong());
+        SnapshotLifecyclePolicy policy = createPolicy(id);
+        long version = randomNonNegativeLong();
+        SnapshotLifecyclePolicyMetadata meta = new SnapshotLifecyclePolicyMetadata(policy, Collections.emptyMap(), version, 1);
+        assertThat(SnapshotLifecycleService.getJobId(meta), equalTo(id + "-" + version));
+    }
+
+    /**
+     * Test new policies getting scheduled correctly, updated policies also being scheduled,
+     * and deleted policies having their schedules cancelled.
+     */
+    public void testPolicyCRUD() throws Exception {
+        ClockMock clock = new ClockMock();
+        final AtomicInteger triggerCount = new AtomicInteger(0);
+        final AtomicReference<Consumer<SchedulerEngine.Event>> trigger = new AtomicReference<>(e -> triggerCount.incrementAndGet());
+        try (ThreadPool threadPool = new TestThreadPool("test");
+             ClusterService clusterService = ClusterServiceUtils.createClusterService(threadPool);
+             SnapshotLifecycleService sls = new SnapshotLifecycleService(Settings.EMPTY,
+                 () -> new FakeSnapshotTask(e -> trigger.get().accept(e)), clusterService, clock)) {
+
+            sls.offMaster();
+            SnapshotLifecycleMetadata snapMeta = new SnapshotLifecycleMetadata(Collections.emptyMap());
+            ClusterState previousState = createState(snapMeta);
+            Map<String, SnapshotLifecyclePolicyMetadata> policies = new HashMap<>();
+
+            SnapshotLifecyclePolicyMetadata policy =
+                new SnapshotLifecyclePolicyMetadata(createPolicy("foo", "*/1 * * * * ?"), // trigger every second
+                    Collections.emptyMap(), 1, 1);
+            policies.put(policy.getPolicy().getId(), policy);
+            snapMeta = new SnapshotLifecycleMetadata(policies);
+            ClusterState state = createState(snapMeta);
+            ClusterChangedEvent event = new ClusterChangedEvent("1", state, previousState);
+            sls.clusterChanged(event);
+
+            // Since the service does not think it is master, it should not be triggered
+            assertThat(triggerCount.get(), equalTo(0));
```

Comment:
I think there might be a race condition that might cause this assert to pass even if the service is running the job when it shouldn't due to so little time passing between the `clusterChanged` call and checking the trigger count.

Reply:
Yeah, this was before I made the triggered function a function. I changed it to be a `fail(...)` if it's triggered during this time so this check is not needed.